server/channels/jobs/batch_report_worker.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/mattermost/mattermost/server/public/shared/request"
+	"github.com/mattermost/mattermost/server/v8/channels/store"
+	"github.com/pkg/errors"
+)
+
+type BatchReportWorkerAppIFace interface {
+	SaveReportChunk(format string, prefix string, count int, reportData []model.ReportableObject) *model.AppError
+	CompileReportChunks(format string, prefix string, numberOfChunks int, headers []string) *model.AppError
+	SendReportToUser(rctx request.CTX, userID string, jobId string, format string) *model.AppError
+	CleanupReportChunks(format string, prefix string, numberOfChunks int) *model.AppError
+}
+
+type BatchReportWorker[T BatchReportWorkerAppIFace] struct {
+	BatchWorker
+	app          T
+	reportFormat string
+	headers      []string
+	getData      func(jobData model.StringMap, app T) ([]model.ReportableObject, model.StringMap, bool, error)
+}
+
+func MakeBatchReportWorker[T BatchReportWorkerAppIFace](
+	jobServer *JobServer,
+	store store.Store,
+	app T,
+	timeBetweenBatches time.Duration,
+	reportFormat string,
+	headers []string,
+	getData func(jobData model.StringMap, app T) ([]model.ReportableObject, model.StringMap, bool, error),
+) model.Worker {
+	worker := &BatchReportWorker[T]{
+		app:          app,
+		reportFormat: reportFormat,
+		headers:      headers,
+		getData:      getData,
+	}
+	worker.BatchWorker = BatchWorker{
+		jobServer:          jobServer,
+		logger:             jobServer.Logger(),
+		store:              store,
+		stop:               make(chan struct{}),
+		stopped:            make(chan bool, 1),
+		jobs:               make(chan model.Job),
+		timeBetweenBatches: timeBetweenBatches,
+		doBatch:            worker.doBatch,
+	}
+	return worker
+}
+
+func (worker *BatchReportWorker[T]) doBatch(rctx *request.Context, job *model.Job) bool {
+	reportData, nextData, done, err := worker.getData(job.Data, worker.app)
+	if err != nil {
+		worker.logger.Error("Worker: Failed to get data for report batch. Exiting", mlog.Err(err))
+		worker.setJobError(worker.logger, job, model.NewAppError("doBatch", model.NoTranslation, nil, "", http.StatusInternalServerError).Wrap(err))
+		return true
+	} else if done {
+		if err = worker.complete(rctx, job); err != nil {
+			worker.logger.Error("Worker: Failed to finish the batch report. Exiting", mlog.Err(err))
+			worker.setJobError(worker.logger, job, model.NewAppError("doBatch", model.NoTranslation, nil, "", http.StatusInternalServerError).Wrap(err))
+		} else {
+			worker.logger.Info("Worker: Report job complete")
+			worker.setJobSuccess(worker.logger, job)
+		}
+
+		return true
+	}
+
+	err = worker.saveData(job, reportData)
+	if err != nil {
+		worker.logger.Error("Worker: Failed to save report batch. Exiting", mlog.Err(err))
+		worker.setJobError(worker.logger, job, model.NewAppError("doBatch", model.NoTranslation, nil, "", http.StatusInternalServerError).Wrap(err))
+		return true
+	}
+
+	for k, v := range nextData {
+		job.Data[k] = v
+	}
```

Comment:
It feels safer to `job.Data = nextData` instead of accidentally preserving "old data" that shouldn't be part of "next data". I realize we write to `job.Data` to increment the `file_count`, but maybe we could return that from `saveData`, assign to `nextData`, and simplify this?

Reply:
I did it to avoid having to force every job to rewrite its own initial metadata (eg. the date range/other filters of the report, which user requested it) on every call of `getData` because I wanted to simplify for those writing jobs that use this architecture.

I'd rather have the jobs have more control over what goes in and out of their metadata instead of trying to be too prescriptive here.

Reply:
> I'd rather have the jobs have more control over what goes in and out of their metadata instead of trying to be too prescriptive here.

Agree with this sentiment, but then wouldn't that mean doing `job.Data = nextData`, and letting jobs have more control? As it stands now, we're assuming they want to preserve old job metadata, but this worker can't know that without being too prescriptive.

Reply:
Hmm I suppose that's true, I didn't consider the case of wanting to have the job completely "unset" a piece of data. I think that's a valid case to change it, so I'll update this.